Refactor string methods for StringArray + return IntegerArray for numeric results

TomAugspurger opened the pull request:
Intended as an alternative to https://github.com/pandas-dev/pandas/pull/29637.

This is a much smaller change. It only changes the codepath for StringDtype.
I think this is OK since someday (well down the road) we'll want to deprecate
the `.str` accessor on object-dtype Series. When we enforce that, we can just
delete the entire old implementation.

The API change is limited to always returning Int64Dtype for numeric outputs, rather than int if there's no NAs and float if there are any.
When BoolArray is done, we'll change that for the boolean-returning ones.

As a side benefit, we get a nice perf boost, since we have deterministic output dtypes we
can skip an object-dtype allocation.

```python
In [2]: s = pd.Series(['a'] * 100_000 + [None], dtype="string")

In [3]: t = s.astype(object)

In [4]: %timeit s.str.count("a")
50.7 ms ± 2.21 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [5]: %timeit t.str.count("a")
62.7 ms ± 742 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

TomAugspurger committed:
take 2

jbrockmendel commented on pandas/core/strings.py:
```diff
@@ -109,9 +114,46 @@ def cat_safe(list_of_columns: List, sep: str):
     return _map(f, arr, na_mask=True, na_value=na_result, dtype=dtype)
 
 
+def _map_ea(f, arr, na_value, dtype):
```
nitpick: `f` -> `func`, annotattion for `dtype` (in particular, is it just np.dtype or can it be EADtype?)

Since this is internal, I'd advocate typing `arr` as having already had extract_array called on it

TomAugspurger committed:
updates

jorisvandenbossche commented on doc/source/user_guide/text.rst:
```diff
@@ -63,7 +63,39 @@ Or ``astype`` after the ``Series`` or ``DataFrame`` is created
+These are places where the behavior of ``StringDtype`` objects differ from
+``object`` dtype
+
+l. For ``StringDtype``, :ref:`String accessor methods<api.series.str>`
```
```suggestion
l. For ``StringDtype``, :ref:`string accessor methods<api.series.str>`
```

jorisvandenbossche commented on pandas/_libs/lib.pyx:
```diff
@@ -2208,9 +2208,13 @@ def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
 @cython.wraparound(False)
-def map_infer_mask(ndarray arr, object f, const uint8_t[:] mask, bint convert=1):
+def map_infer_mask(ndarray arr, object f, const uint8_t[:] mask, bint convert=1,
+                   object na_value=_no_default, object dtype=object):
```
Can you update the docstring here?

jorisvandenbossche commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,48 @@ def cat_safe(list_of_columns: List, sep: str):
 def _na_map(f, arr, na_result=np.nan, dtype=object):
     # should really _check_ for NA
+    if is_extension_array_dtype(arr.dtype):
+        arr = extract_array(arr, extract_numpy=True)
```
```suggestion
        arr = extract_array(arr)
```

? 
`_map_ea` should always get a extension array, never a numpy array

jorisvandenbossche commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,48 @@ def cat_safe(list_of_columns: List, sep: str):
+        return IntegerArray(result, mask)
+
+    elif is_string_dtype(dtype) and not is_object_dtype(dtype):
+        result = lib.map_infer_mask(arr, func, mask.view("uint8"), na_value=na_value)
```
convert=False here as well? (the array is object dtype, and should stay object?)

jorisvandenbossche commented:
> As a side benefit, we get a nice perf boost

Maybe a copy paste error, but the example doesn't show much performance boost :) (less than a percent)

jorisvandenbossche commented:
This looks good to me! I suppose this is simpler than your other PR

jbrockmendel commented:
Is this time-sensitive or can i wait and look at it on Monday?

TomAugspurger commented:
No rush on this.

I did fail on the copy paste. The speed up is about 15% IIRC.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into masked-string-2

TomAugspurger committed:
updates

WillAyd commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,51 @@ def cat_safe(list_of_columns: List, sep: str):
 
 
+def _map_ea(
+    func: Callable, arr: "StringArray", na_value: Any, dtype: Dtype
```
Can you annotate what the `Callable` accepts and returns? I think very helpful to identify, especially for new development

jorisvandenbossche replied:
I suppose it always accepts a string, but the return value is variable (string, int, boolean, list, ...)

WillAyd commented on pandas/tests/test_strings.py:
```diff
@@ -3502,8 +3514,29 @@ def test_string_array(any_string_method):
+)
+def test_string_array_numeric_integer_array(method, args, expected):
+    s = Series(["aba", None], dtype="string")
+    result = getattr(s.str, method)(*args)
```
Since all the same could simplify this and not even parametrize `args`

jreback commented on doc/source/user_guide/text.rst:
```diff
@@ -63,7 +63,39 @@ Or ``astype`` after the ``Series`` or ``DataFrame`` is created
+   that return **numeric** output will always return a nullable integer dtype,
+   rather either int or float dtype, depending on the presence of NA values.
+
+   .. ipython:: python
```
can you create & show s in an earlier ipython block

jreback commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,51 @@ def cat_safe(list_of_columns: List, sep: str):
+        )
+        return StringArray(result)
+    # TODO: BooleanArray
+    else:
```
what actually hits this branch? IIUC, dtype is explicity passed as either Int64 or string, *never* object by the string routines, albeit you have a default of object.

TomAugspurger replied:
Right now, we hit this with the return types

1. bool (e.g. contains)
2. bytes (`encode`)
3. object -> list (findall, which returns an ndarray[list]
4. object -> Any (`get`)

We'll be able to avoid this for bool. Won't be able to avoid it for nested data or `.get`.

jreback replied:
can you indicate this with a comment.

jreback commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,51 @@ def cat_safe(list_of_columns: List, sep: str):
 
 def _na_map(f, arr, na_result=np.nan, dtype=object):
     # should really _check_ for NA
+    if is_extension_array_dtype(arr.dtype):
```
can you sync the signatures up with _map, also rename _map -> _map_arraylike (or similar)

TomAugspurger replied:
Sorry, don't follow this. I think they do match, aside from @jbrockmendel's request to call it `func` rather than `f`.

I guess `na_map` calls it `na_result` while `_map` calls it `na_value`.

jreback replied:
I am really -1 on 2 different branches here. If they have *exactly* the same signature a little less negative. again I would rename _map to be more informative here

jorisvandenbossche replied:
What do you mean with "-1 on 2 different branches here" ? 
The whole purpose of this PR is to add a separate branch in case of StringArray (because we can be more efficient and want to be more specific in the result dtype)

TomAugspurger replied:
That's why my initial PR was so much more complex, since it tried to handle both cases similarly. I think that was more complex than this.

As Joris says, the main point of divergence is that for StringArray we usually know the result dtype exactly. It doesn't depend on the presence of NAs. Additionally,

1. We're still using `map_infer_dtype` for both, so the core implementation is the same.
2. We'll eventually deprecate `.str` on object-dtype, so we will end up with just this implementation.

jreback replied:
ok, at the very least these signatures for what you call _map and _stringarray_map should be *exactly* the same.
and _map -> _map_object and _stringarray_map -> _map_stringarray.

I think this is crucial for not adding technical debt.

jbrockmendel replied:
@jreback is your suggestion to add `na_mask` to the _stringarray_map signature and have it just not be used?  I think this relates to the "this should be a StringArray method" discussion

TomAugspurger replied:
`_map` and `_na_map` already have inconsistent signatures. I'm not sure why it's that way on master, but I'm a bit against adding unused arguments in this case.

What's the technical debt we're adding here? By definition, we need to handle StringArray differently, since its result type doesn't depend on the presence of NAs.

jorisvandenbossche replied:
> _map and _na_map already have inconsistent signatures. I'm not sure why it's that way on master, but I'm a bit against adding unused arguments in this case.

And there is also a good reason for that, as `_map` has an additional argument `na_mask` that is used internally in `_map` (for a recursive call). 
I think refactoring `_map` is outside of the scope of this PR.

jreback replied:
can you create an issue to clean this up (_map and _na_map), and/or refactor of this, post this PR.

TomAugspurger committed:
type callable

TomAugspurger committed:
wip

TomAugspurger committed:
more dtypes

TomAugspurger commented:
Just the clipboard failure: https://github.com/pandas-dev/pandas/issues/29676

jreback commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,51 @@ def cat_safe(list_of_columns: List, sep: str):
+    assert isinstance(arr, StringArray)
+    arr = arr._ndarray
+
+    if is_integer_dtype(dtype):
```
pls add some more comments on what is being branched on (the return dtype) and why we are doing this.

jreback commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,51 @@ def cat_safe(list_of_columns: List, sep: str):
+        return lib.map_infer_mask(arr, func, mask.view("uint8"))
+
+
 def _map(f, arr, na_mask=False, na_value=np.nan, dtype=object):
```
This approach is creating a bunch of technical debt here. ok for now, but we need a path forward to combine these two approaches, one for EA, one for object.

TomAugspurger replied:
That's why my initial PR was so much more complex, since it tried to handle both similarly.

But I think it's not so bad, because

1. We're still using `map_infer_dtype` for both, so the core implementation is the same.
2. We'll eventually deprecate `.str` on object-dtype, so we will end up with just this implementation.

TomAugspurger replied:
Let's have this convo in https://github.com/pandas-dev/pandas/pull/29640#discussion_r347137430

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into masked-string-2

TomAugspurger committed:
doc

TomAugspurger commented:
CI failure is from https://github.com/pandas-dev/pandas/issues/29514

jbrockmendel commented on doc/source/user_guide/text.rst:
```diff
@@ -63,7 +63,40 @@ Or ``astype`` after the ``Series`` or ``DataFrame`` is created
+
+l. For ``StringDtype``, :ref:`string accessor methods<api.series.str>`
+   that return **numeric** output will always return a nullable integer dtype,
+   rather either int or float dtype, depending on the presence of NA values.
```
"rather" --> "rather than"?

jbrockmendel commented on pandas/_libs/lib.pyx:
```diff
@@ -2208,31 +2208,47 @@ def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
     ----------
     arr : ndarray
     f : function
+    mask : ndarray
```
`ndarray[uint8]`

TomAugspurger replied:
How do we handle dtypes in docstrings? I didn't think numpy supported that yet.

jbrockmendel replied:
this is the pattern we use in the cython docstrings, but we're not super-consistent about it.  looking forward to if/when numpy supports that kind of annotation

jbrockmendel commented on pandas/_libs/lib.pyx:
```diff
@@ -2208,31 +2208,47 @@ def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
+        uint8 dtype ndarray indicating values not to apply `f` to.
+    convert : bool, default True
+        Whether to call `maybe_convert_objects` on the resulting ndarray
+    na_value : Any, optional
```
scalar?

TomAugspurger replied:
No, I don't think so. These can return nested objects (e.g. `.str.get`)

jbrockmendel commented on pandas/_libs/lib.pyx:
```diff
@@ -2208,31 +2208,47 @@ def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
+    na_value : Any, optional
+        The result value to use for masked values. By default, the
+        input value is used
+    dtype : type
```
type or np.dtype?

jbrockmendel commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,77 @@ def cat_safe(list_of_columns: List, sep: str):
 
 def _na_map(f, arr, na_result=np.nan, dtype=object):
     # should really _check_ for NA
+    if is_extension_array_dtype(arr.dtype):
```
comment that this is sufficient to ensure StringDType (not just EA dtype)?

jbrockmendel commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,77 @@ def cat_safe(list_of_columns: List, sep: str):
 
 
+def _stringarray_map(
+    func: Callable[[str], Any], arr: "StringArray", na_value: Any, dtype: Dtype
```
na_value restricted to scalar?

TomAugspurger replied:
No, I don't think so.

jbrockmendel commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,77 @@ def cat_safe(list_of_columns: List, sep: str):
+    na_value : Any
+        The value to use for missing values. By default, this is
+        the original value (NA).
+    dtype : Dtype
```
Elsewhere we say "np.dtype or ExtensionDtype".  Is this the new policy?

TomAugspurger replied:
Not sure if we have a policy. If it matters, this is an internal docstring, so I'm OK to use things from `pandas._typing` that we wouldn't have in a public docstring yet.

jbrockmendel commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,77 @@ def cat_safe(list_of_columns: List, sep: str):
+    mask = isna(arr)
+
+    assert isinstance(arr, StringArray)
+    arr = arr._ndarray
```
if this is depending on implementation details of StringArray, then shouldnt it be a StringArray method?

TomAugspurger replied:
Possibly, but need to balance that against the consistency with the rest of the file. IMO, keeping it near the other string methods makes the most sense.

jorisvandenbossche replied:
It's now using `np.asarray`, so not depending on implementation details of StringArray, so marking this comment as resolved.

TomAugspurger committed:
fixups

TomAugspurger commented:
Thanks for the review. Addressed all the comments I think.

After this, I'll have a small change at https://github.com/pandas-dev/pandas/pull/29597 to get it using the new methods (there's some no unnecessary changes in https://github.com/pandas-dev/pandas/pull/29597 that need to be removed).

TomAugspurger committed:
use asarray

TomAugspurger committed:
use asarray 2

TomAugspurger commented:
Just the clipboard CI failure again.

TomAugspurger commented:
@jreback thoughts on the conversation around the signatures? I think this is slightly blocking the pd.NA PR now. 

> On Nov 19, 2019, at 06:01, Joris Van den Bossche <notifications@github.com> wrote:
> 
> ﻿
> @jorisvandenbossche approved this pull request.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

jreback commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -63,7 +63,7 @@ Previously, strings were typically stored in object-dtype NumPy arrays.
    and parts of the API may change without warning.
 
-The text extension type solves several issues with object-dtype NumPy arrays:
+The ``'string'`` extension type solves several issues with object-dtype NumPy arrays:
```
does this need any updating an/or reference to your new section that you added in text.rst?

jorisvandenbossche replied:
@TomAugspurger you can probably update the sentence "The usual string accessor methods work. Where appropriate, the return type of the Series or columns of a DataFrame will also have string dtype." 20 lines below this line, to include that it can also return IntegerDtype in certain cases.

jreback commented on pandas/core/strings.py:
```diff
@@ -109,9 +118,78 @@ def cat_safe(list_of_columns: List, sep: str):
+    if is_extension_array_dtype(arr.dtype):
+        # just StringDtype
+        arr = extract_array(arr)
+        return _stringarray_map(f, arr, na_value=na_result, dtype=dtype)
```
I would rename these as indicated above.

jorisvandenbossche replied:
> I would rename these as indicated above.

Can you be explicit in what exact names you want to propose? (I am getting a bit lost in all the comments) 
Tom already renamed `_ea_map` to `_stringarray_map`

jreback replied:
right we should change _map -> _object_map (though I actually prefer the opposite, _map_object and _map_stringarray), but either is ok.

TomAugspurger replied:
Thanks, make sense. I renamed to `_map_object` and `_map_stringarray`

jreback reviewed (changes requested):
very minor comments, rebase and ping on green.

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into masked-string-2

TomAugspurger committed:
fixup

TomAugspurger commented:
Merged master, and writing up an issue for the followup.

TomAugspurger committed:
rename

TomAugspurger commented:
https://github.com/pandas-dev/pandas/issues/29710 for followup.

TomAugspurger commented:
Thanks all!
